### ISSUE TITLE
`manual_unwrap_or_default` does not take into account mutable code in the else branch

### ISSUE THREAD

**Issue by orlp** [rust-lang/rust-clippy#12580]
### Summary

Nightly clippy with `manual_unwrap_or_default` seems to ignore mutable code in the else branch.

### Lint Name

manual_unwrap_or_default

### Reproducer

I tried this code:

```rust
fn foo(num_missing: &mut u32, opt: Option<u32>) -> u32 {
    if let Some(r) = opt {
        r
    } else {
        *num_missing += 1;
        0
    }
}
```

I saw this happen:

```rust
warning: if let can be simplified with `.unwrap_or_default()`
 --> src/lib.rs:2:5
  |
2 | /     if let Some(r) = opt {
3 | |         r
4 | |     } else {
5 | |         *num_missing += 1;
6 | |         0
7 | |     }
  | |_____^ help: replace it with: `opt.unwrap_or_default()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_unwrap_or_default
  = note: `#[warn(clippy::manual_unwrap_or_default)]` on by default
```

I expected to see this happen:

No lint emitted as it's not applicable.

### Version

```text
rustc 1.79.0-nightly (c9f8f3438 2024-03-27)
binary: rustc
commit-hash: c9f8f3438a8134a413aa5d4903e0196e44e37bbc
commit-date: 2024-03-27
host: x86_64-unknown-linux-gnu
release: 1.79.0-nightly
LLVM version: 18.1.2
```
